docs/adr/0004-2021-03-03-frontend-e2e-tests.md

```diff
@@ -0,0 +1,65 @@
+# Fontend end-to-end testing
+
+## Status
+
+Accepted
+
+## Context
+
+The [frontend application](https://github.com/mcagov/beacons-webapp) currently does not have end-to-end tests, and we feel that adding them would help development of the app.
+
+## Decision
+
+We will use [Cypress](https://docs.cypress.io/guides/overview/why-cypress.html#In-a-nutshell) to write end-to-end tests for the frontend app in the [frontend repository](https://github.com/mcagov/beacons-webapp).
+
+These tests will be used to check the functionality of the application and its pages.
+
+These tests are `end-to-end` within the frontend app - we would mock any API calls from the frontend.
```

Comment:
Hmm this isn't my understanding of `end-to-end`...  I found this diagram really helpful:

![image](https://user-images.githubusercontent.com/54271433/109962102-b705dd00-7ce2-11eb-91d7-3c13e265014a.png)

Along with these definitions:

![image](https://user-images.githubusercontent.com/54271433/109962138-bff6ae80-7ce2-11eb-8e36-8e7d23fa355f.png)

(Though there is no universal definition of "What is an X test?" etc.)

So if we're mocking the API then I think the type of testing we're doing with Cypress in the frontend only is  `integration` testing, the integration being of View and Controller.

But I'm really not sure myself, or even if that distinction is useful or not 😂

Reply:
I think I would also name these as `integration` tests if we're mocking out the API/database layer. To be honest so long as we're consistent within the project and, as a team, we know what we mean when we say unit/integration/e2e tests.  For the beacons project I see this as:

- Unit tests - isolated units tested in both the webapp and API repositories
- Integration tests - tests between system boundaries.  API = API service and Database, Webapp = user/browser and webapp
- End-to-End tests = Testing end-to-end flow = webapp, service, database

Reply:
Thank you both! This definitely helps my understanding of when to use those terms! 🙏